### PR TITLE
Fix frozen pages on shared browser connections

### PR DESCRIPTION
**Pull request by ankur22**
## What?

When multiple VUs share the same remote Chrome instance, each VU gets notified about every new page opened in that browser, including pages opened by other VUs. With this change, a VU only keeps hold of the pages it opened itself. A page that belongs to another VU is immediately released, so its owner can work with it right away.

Made up of these core changes:

* Fix a race issue.
* Ensure we not only resume (now awaited) but also detach from targets that are not part of the current VU's target (owned by another VU).
* Correct filtering of foreign targets.
* detach session of foreign target (don't just early return) if the target slips through to the second gate (isAttachedPageValid).

All of these help:

* prevent the blocking of the target which is owned by another VU.
* align the fix for all edge cases where this might happen.
* align with Playwright.

## Why?

When working with remote Chrome instances, let's say 2 of them, and a test that requires 50 VUs, we found that the browser module would indeed start all 50 VUs, but only 2 would be able to work with the Chrome browsers. All the others had to wait until a running iteration ended before one of them could take the freed spot, so the test effectively ran 2 VUs at a time instead of 50.

The cause: a new page always starts out paused, and it only starts running once every VU connected to that browser gives it the green light. A VU that was still setting up its own session would take hold of pages belonging to other VUs and never give that green light. Those pages stayed paused until the holding VU finished its iteration and disconnected.

This can only happen when VUs share a browser instance (remote browsers via `K6_BROWSER_WS_URL`). When every VU launches its own browser, which is what local runs do, nothing is shared, so this never showed up locally.

Playwright follows the same rule we adopt here: every target it gets attached to is either adopted and resumed ([crPage.ts#L549](https://github.com/microsoft/playwright/blob/2b1e2f67e84904b816d0ca0c8f2dcad5588d82e4/packages/playwright-core/src/server/chromium/crPage.ts#L549)), or resumed and detached from ([crConnection.ts#L176-L185](https://github.com/microsoft/playwright/blob/2b1e2f67e84904b816d0ca0c8f2dcad5588d82e4/packages/playwright-core/src/server/chromium/crConnection.ts#L176-L185), which notes that detaching alone should release a paused target but a backend bug makes the explicit resume necessary). No target is ever silently held, and this change applies the same rule at every place k6 declines a target.

The new test reproduces the freeze on master and passes in under a second with the fix.

Tested with `-race` on the browser packages, no new lint issues. Full `make check` not run yet (draft).

## Checklist

- [x] I have performed a self-review of my code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have added tests for my changes.
- [x] I have run linter and tests locally (`make check`) and all pass.

Replaces #6276: same change rebuilt so the release mechanism (resume awaited, then detach) exists from the first commit, plus a regression test that the detach is only sent after the browser acknowledges the resume. Each commit builds, lints, and passes tests independently.

### CODE REVIEW

File: internal/js/modules/k6/browser/common/browser.go [hunk 1]
```diff
@@ -322,7 +323,7 @@ func (b *Browser) onAttachedToTarget(ev *target.EventAttachedToTarget) error {
 	if err != nil && b.isPageAttachmentErrorIgnorable(ev, session, err) {
 		if b.closing.Load() {
 			b.logger.Debugf("Browser:onAttachedToTarget", "new page failed; browser is closing: sid:%v", ev.SessionID)
-			detachSession(b.browserCtx, session)
+			detachSession(session)
```
Comment (janHildebrandt98) [inline internal/js/modules/k6/browser/common/browser.go]: Should this `detachSession(session)` fire even when `b.closing` is false? `isPageAttachmentErrorIgnorable` can return true from a canceled `vuCtx` too, and `b.browserCtx` (what the `Connection` actually lives on) isn't derived from `vuCtx`.

File: internal/js/modules/k6/browser/common/frame_session.go [hunk 2]
```diff
@@ -969,7 +969,7 @@ func (fs *FrameSession) onAttachedToTarget(event *target.EventAttachedToTarget)
 	default:
 		fs.logger.Debugf("FrameSession:onAttachedToTarget",
 			"unsupported target type %q sid:%v", ti.Type, session.ID())
-		detachSession(fs.teardownCtx, session)
+		detachSession(session)
```
Comment (janHildebrandt98) [inline internal/js/modules/k6/browser/common/frame_session.go]: `detachSession` went from a synchronous call bound to `fs.teardownCtx` to an async goroutine bound to the connection's ctx and `DefaultTimeout` (same for the other three call sites at 1023, 1062, 1076). Makes sense given the ordering fix elsewhere in the PR, but none of the new tests exercise these `FrameSession` paths (`unsupported target type`, `page is closing` for iframe/worker). Worth a quick test here too, or are you comfortable relying on the connection/browser-level tests as coverage since the underlying `detachSession` is shared?

File: internal/js/modules/k6/browser/common/connection.go [hunk 3]
```diff
@@ -445,29 +448,35 @@ func (c *Connection) recvLoop() {
...
+// is browser-level: the session goes in the params, not the message's
+// session ID.
+func detachSession(session *Session) {
+	go func() {
+		c := session.conn
+		ctx, cancel := context.WithTimeout(c.ctx, DefaultTimeout)
+		defer cancel()
+		_ = session.Execute(ctx, cdpruntime.CommandRunIfWaitingForDebugger, nil, nil)
+
+		buf, err := jsonv2.Marshal(&target.DetachFromTargetParams{SessionID: session.id}, defaultJSONV2Options)
```
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: `Target.detachFromTarget` addressed at the browser level (no envelope session ID) only works when the session was attached by the root/browser session. Four call sites in `frame_session.go` (this one plus 1023, 1062, 1076) hand `detachSession` a session that a *page* session auto-attached, not the browser. Real Chrome (131 and 151) returns `-32602 No session with given id` for that. Playwright sends this on the parent session instead, for exactly this reason (`crConnection.ts`'s `CRSession.detach`).

Not a regression in outcome, the old self-addressed form failed here too, just with a different error. Severity's bounded: three of the four sites only fire while the page is already closing, and Chrome detaches every child session when the page target closes anyway, so the residue is milliseconds. The one that isn't bounded is the `default:` case in `frame_session.go` (unsupported target type, e.g. a live service worker). That one can leave a session attached on an otherwise-healthy page, with nothing logged anywhere, since the error reply has no session ID and misses `recvLoop`'s own `No session with given id` handling.

One more small thing this changes: pre-PR, that failed detach came back addressed to the session, which tripped the existing cleanup path and closed the local `Session` object anyway. Now nothing cleans it up until the page closes.

File: internal/js/modules/k6/browser/common/connection.go [hunk 4]
```diff
@@ -363,21 +366,21 @@ func (c *Connection) recvLoop() {
 			eva := ev.(*target.EventAttachedToTarget) //nolint:forcetypeassert
 			sid, tid := eva.SessionID, eva.TargetInfo.TargetID
 
+			c.sessionsMu.Lock()
+			session := NewSession(c.ctx, c, sid, tid, c.logger, c.msgIDGen)
```
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Since this creates the `Session` before the reject filter runs now, every foreign target that gets rejected here builds a full session, including its own event-emitter goroutine, which only exits on the connection's own ctx, not on `session.close()`. So each rejected target parks one live goroutine for the life of the connection, not just until its map entry gets cleaned up.

It's the same lifetime gap accepted sessions already have, so not a new leak shape, and it's bounded to one k6 iteration (`Browser.Close()` cancels the connection ctx and reaps them). In the 50-VU shared-browser case this PR is fixing, that's one extra parked goroutine per page anyone ever opened in that browser, not per VU. Still, worth a look, a per-session cancelable context would close it.

File: internal/js/modules/k6/browser/common/browser.go [hunk 5]
```diff
@@ -160,12 +160,16 @@ func (b *Browser) connect() error {
...
-	// We don't need to lock this because `connect()` is called only in NewBrowser
-	b.defaultContext, err = NewBrowserContext(b.vuCtx, b, "", DefaultBrowserContextOptions(), b.logger)
+	defaultContext, err := NewBrowserContext(b.vuCtx, b, "", DefaultBrowserContextOptions(), b.logger)
 	if err != nil {
 		return fmt.Errorf("browser connect: %w", err)
 	}
-	b.runOnClose = append(b.runOnClose, b.defaultContext.cleanup)
+	// The connection's recvLoop reads defaultContext through
+	// connectionOnAttachedToTarget, so publish it under contextMu.
+	b.contextMu.Lock()
```
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/browser.go]: Since you're locking `b.defaultContext`'s publish here, is `disposeContext`'s `b.context = nil` an oversight? It's a plain unlocked write (not shown in this diff), and `connectionOnAttachedToTarget` reads `b.context` under this same lock a few lines down. `-race` catches the collision: closing a `BrowserContext` while an attach event lands on a shared browser races on `b.context`, reproduced against a live Chrome and `recvLoop`.

Looks pre-existing (blame points to 2023), not from this PR, but since you're already fixing the exact sibling of it in this same commit, might be worth grabbing here too. `NewContext`, `CloseContext`, and `Context` further down also read/write `b.context` unlocked, if you go down this road.

File: internal/js/modules/k6/browser/common/connection.go [hunk 6]
```diff
@@ -445,29 +448,35 @@ func (c *Connection) recvLoop() {
...
+// detachSession resumes a rejected target's session, awaits the response,
+// then detaches from it. Detaching alone should release the target, but the
+// browser has a bug where a detached target can stay paused, so the resume
+// must land first — the same workaround as Playwright's CRSession.detach
+// (crConnection.ts). The wait is bounded so an unresponsive target is still
+// detached from. Best-effort: errors are not returned. Target.detachFromTarget
+// is browser-level: the session goes in the params, not the message's
+// session ID.
+func detachSession(session *Session) {
+	go func() {
```
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Worth a one-line comment that this goroutine isn't just for throughput: `detachSession` is called from `recvLoop` itself now, and it waits on a reply only `recvLoop` can deliver, so without it you'd deadlock the one goroutine reading the websocket for up to 30 seconds per rejected target. The old `stopWaitingForDebugger` was safe to call inline, so this would be an easy trap for someone to "simplify" back later without that context.

File: internal/js/modules/k6/browser/common/frame_session.go (hunk 2 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/frame_session.go]: Now that `detachSession` doesn't take a ctx, is `FrameSession.teardownCtx` still doing anything? Looks like the field, the `NewFrameSession` param, and passing it down to child iframe sessions are all that's left of it, and nothing reads it for its `Done()` anymore. `Page.teardownCtx` still has a real use (`page.go:976`), it's just this copy on `FrameSession` that's dead now.

File: internal/js/modules/k6/browser/common/browser.go [hunk 7]
```diff
@@ -266,23 +270,16 @@ func (b *Browser) initEvents() error {
 }
 
 // connectionOnAttachedToTarget is called when Connection receives an attachedToTarget
-// event. Returning false will stop the event from being processed by the connection.
+// event. Returning false makes the connection release the target instead of
+// adopting it. Targets from the connection's own browser context and from
+// the default browser context are accepted.
```
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/browser.go]: Is this comment accurate against real Chrome? `b.defaultContext.id` is always `""` (set via `NewBrowserContext(..., "", ...)`), but Chrome reports a non-empty `browserContextId` for every target, default profile included, so `b.defaultContext.id == eva.TargetInfo.BrowserContextID` can't be true outside the unit test, which fabricates an empty ID. Not a bug, the effective rule is "accept only my own context," which is what you want, just the comment describes a branch that can't fire in production.

File: internal/js/modules/k6/browser/common/connection.go (hunk 3 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: `Target.detachFromTarget` on the browser session only works for sessions the browser auto-attached, and the iframe and worker sessions coming from `frame_session.go` are attached to the page, so real Chrome answers `-32602 No session with given id`. Playwright sends it on the parent session for the same reason (`CRSession.detach` in `crConnection.ts`). The `default:` case then leaves a live service worker's session attached, and we never log it.

File: internal/js/modules/k6/browser/common/connection.go (hunk 4 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Building the `Session` before the reject filter means every rejected target now leaves its event-emitter goroutine running: it exits on the connection's ctx, not on `session.close()`. Would a per-session cancelable context be worth it here?

File: internal/js/modules/k6/browser/common/browser.go (hunk 5 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/browser.go]: Since you're locking `b.defaultContext`'s publish here, is `disposeContext`'s `b.context = nil` an oversight? It writes without the lock, and `connectionOnAttachedToTarget` reads `b.context` under this lock a few lines down, so closing a context while a target attaches races. `-race` against real Chrome flags it, and it's been there since 2023, but you're fixing its sibling right here.

File: internal/js/modules/k6/browser/common/connection.go (hunk 6 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Worth a one-line comment that this goroutine isn't just style? `detachSession` runs from `recvLoop` now and waits for a reply only `recvLoop` can deliver, so dropping it deadlocks the one goroutine reading the websocket for up to 30s per rejected target.

File: internal/js/modules/k6/browser/common/frame_session.go (hunk 2 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/frame_session.go]: Now that `detachSession` doesn't take a ctx, does `FrameSession.teardownCtx` still do anything? The field and the `NewFrameSession` param look like they only exist to pass it to child frame sessions now. `Page.teardownCtx` still has a real use at `page.go:976`, just not this copy.

File: internal/js/modules/k6/browser/common/browser.go (hunk 7 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/browser.go]: Is this comment accurate? `b.defaultContext.id` is always `""`, but real Chrome reports a non-empty `browserContextId` even for default-profile targets, so this branch only fires in the unit test, which fabricates an empty id.

File: internal/js/modules/k6/browser/common/connection.go (hunk 3 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Should this address the session rather than the browser, the way Playwright's `CRSession.detach` in `crConnection.ts` does? `Target.detachFromTarget` sent on the browser session only reaches targets the browser auto-attached, so the iframe and worker sessions coming from `frame_session.go` come back as `-32602 No session with given id`. Three of those four call sites fire while the page is closing and Chrome detaches the children anyway, but the `default:` case can leave a live service worker's session attached and we drop the error. Could that one at least log the failure?

File: internal/js/modules/k6/browser/common/connection.go (hunk 4 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Can `NewSession` move below the reject check, or does the reject path need the session for something? Building it first leaves every rejected target with its `BaseEventEmitter` sync goroutine parked until the connection's ctx ends, since `session.close()` doesn't stop it. Pre-PR a rejected target created none, and it's one goroutine per rejected target for the browser's lifetime, so nothing urgent.

File: internal/js/modules/k6/browser/common/browser.go (hunk 5 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/browser.go]: Is `disposeContext`'s `b.context = nil` an oversight now that this publish takes the lock? It writes without `contextMu` while `connectionOnAttachedToTarget` reads `b.context` under `RLock` a few lines down, so closing a context while a target attaches trips `-race` against real Chrome. It predates this PR, but a shared browser makes it much easier to hit, so could you take the same lock there while you're in here?

File: internal/js/modules/k6/browser/common/connection.go (hunk 6 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Could you add a line here saying this goroutine is required rather than stylistic? `detachSession` runs from `recvLoop` now and waits on a reply only `recvLoop` can deliver, so dropping the `go` would block the single websocket reader for up to 30s per rejected target. Without a note it looks like noise someone can remove later.

File: internal/js/modules/k6/browser/common/frame_session.go (hunk 2 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/frame_session.go]: Does `FrameSession.teardownCtx` still do anything now that `detachSession` takes no ctx? Nothing reads it except to pass it down to child frame sessions, and removing the field, the `NewFrameSession` param and that plumbing still compiles. `Page.teardownCtx` keeps a real use at `page.go:976`, so it's only this copy; can we drop it?

File: internal/js/modules/k6/browser/common/browser.go (hunk 7 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/browser.go]: Can we reword this comment? `b.defaultContext.id` is always `""`, while real Chrome reports a non-empty `browserContextId` even for default-profile targets, so the default-context half only fires in the unit test that fabricates an empty id. Accepting just the connection's own context is what we want anyway, the comment only claims more than the code does. Or have you seen Chrome send an empty id somewhere?

File: internal/js/modules/k6/browser/common/connection.go (hunk 3 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: This sends `Target.detachFromTarget` with no session id in the message, so Chrome treats it as a browser-level command. That works for a session the browser attached directly, but the iframe sessions coming out of `attachIFrameToTarget` were attached through the page's session, not the browser's, so Chrome answers `-32602 No session with given id`. Doesn't cause trouble today, since this only runs while the page's already closing and Chrome tears the iframe down anyway, but shouldn't this go out through the session itself, like Playwright's `CRSession.detach` does?

File: internal/js/modules/k6/browser/common/connection.go (hunk 4 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go]: Could `NewSession` move below the reject check here? Building it first leaves the rejected target's session with nothing ever calling `session.close()` on it, so its `BaseEventEmitter` keeps syncing for as long as the connection stays open. Before this, a rejected target created zero goroutines. Now it's one per rejected target, for the life of the browser.

File: internal/js/modules/k6/browser/common/connection.go [hunk 8]
```diff
@@ -445,29 +448,35 @@ func (c *Connection) recvLoop() {
...
+		if err != nil {
+			c.logger.Errorf("Connection:detachSession", "sid:%v wsURL:%q, err:%v", session.id, c.wsURL, err)
+			return
+		}
+		msg := &cdproto.Message{
+			ID:     c.msgIDGen.newID(),
+			Method: cdproto.MethodType(target.CommandDetachFromTarget),
+			Params: buf,
+		}
+		if err := c.send(c.ctx, msg, nil, nil); err != nil {
```
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go:504]: Shouldn't this go out through the session itself? When a target attaches to a page (like an iframe), Chrome returns an error when we send it this message: it can't find the target because it would belong to a page.

File: internal/js/modules/k6/browser/common/connection.go (hunk 4 above)
Comment (inancgumus) [inline internal/js/modules/k6/browser/common/connection.go:399]: This leaks goroutines since we build the session before rejecting targets. Their sessions might stay open.

**Review by inancgumus — DISMISSED** [review]
Nice fix 🎉

**Review by janHildebrandt98 — DISMISSED** [review]

File: internal/js/modules/k6/browser/common/browser.go (hunk 1 above)
Comment (ankur22) [reply internal/js/modules/k6/browser/common/browser.go]: Thanks, fixed in 9cd7ffa40

File: internal/js/modules/k6/browser/common/connection.go (hunk 8 above)
Comment (ankur22) [reply internal/js/modules/k6/browser/common/connection.go:504]: Good spot!

I tested this against real Chrome (cross-origin iframe target) with both methods (browser-level and routed via the page's session) and both fail identically ("No session with given id").

Instead, `detachSession` no longer relies on the detach succeeding. It cleans up its local session regardless of the outcome, so this failure mode is now harmless rather than something to route around. Fixed in 217d7f6e2.

File: internal/js/modules/k6/browser/common/connection.go (hunk 4 above)
Comment (ankur22) [reply internal/js/modules/k6/browser/common/connection.go:399]: Linked to the other comment.

Yep, confirmed that the session was left open until the browser's own detach event, which isn't guaranteed to arrive. detachSession now closes the session locally right after its resume-and-detach attempt, regardless of whether it succeeded.

However, that introduced a race. `recvLoop` can hold a session pointer it's about to deliver to while this same cleanup runs concurrently, and the dispatch had no escape if the session's reader was already gone. This is now fixed alongside this by extracting the dispatch into `deliverToSession` with a `<-session.done` case, with tests proving both the race and the leak fix. Fixed in 1f8f3b499 and 217d7f6e2.

File: internal/js/modules/k6/browser/common/connection.go [hunk 9]
```diff
@@ -332,7 +332,37 @@ func (c *Connection) findTargetIDForLog(id target.SessionID) target.ID {
...
+// just because one session closed - stalling recvLoop, and with it every
+// other session on the connection.
+//
 //nolint:funlen,gocognit,cyclop
+func (c *Connection) deliverToSession(session *Session, msg *cdproto.Message) (stop bool) {
+	select {
+	case session.readCh <- msg:
+	case <-session.done:
+		c.logger.Debugf("Connection:deliverToSession:<-session.done", "sid:%v tid:%v wsURL:%q",
+			session.id, session.targetID, c.wsURL)
```
Comment (ankur22) [inline internal/js/modules/k6/browser/common/connection.go:351]: This is the new bit

**Review by inancgumus — APPROVED** [review]

**Review by janHildebrandt98 — APPROVED** [review]
